boards: arduino_nicla_vision: add initial support for the camera system

uLipe opened the pull request:
This PR adds the support of the camera subsystem for the Arduino Nicla Vision, in order  to make it to work this PR covered:

* Enable camera subsystem and the GC2145 on the device tree
* Update the GC2145 video sensor to support lower resolutions
* Add support this board as supported in  the video/capture samples

uLipe committed:
boards: arduino_nicla_vision: add camera subsystem

initial support on device tree

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>

josuah commented on drivers/video/gc2145.c:
```diff
@@ -972,8 +972,7 @@ static uint8_t gc2145_check_connection(const struct device *dev)
 	if ((reg_ver_val != GC2145_REV_VAL) || (reg_pid_val != GC2145_PID_VAL)) {
-		LOG_ERR("failed to detect GC2145 pid: 0x%x rev: 0x%x", reg_pid_val, reg_ver_val);
-		return -EIO;
+		LOG_WRN("Unexpected GC2145 pid: 0x%x and rev: 0x%x", reg_pid_val, reg_ver_val);
```
I suppose this is some issue you faced where the sensor returns something unexpected?
Is it consistent i.e. always the same ID for a given board or does it come with a random number?
Or is it that you tried with different boards and always had a different ID?
The answer to this would be valuable as a comment for the next folk trying to debug the sensor.
Thank you!

uLipe replied:
I'm wondering how to do that, because whay I found is, the pid are always the same, but the revision can change, I think it would be better do push a warning here, and add to the binding a non-required id + revision pair properties.

josuah replied:
Interesting option: then if in a product a wrong/unexpected GalaxyCore part gets delivered, then it would be possible to immediately discover it through this error/warning

An alternative would be to check (`LOG_ERR(); return ret;`) the PID and only warn if the Rev ID changes.

josuah commented on boards/arduino/nicla_vision/arduino_nicla_vision_stm32h747xx_m7.dts:
```diff
@@ -192,3 +186,81 @@ zephyr_udc0: &usbotg_hs {
+	pinctrl-names = "default";
+	status = "okay";
+
+	sensor = <&gc2145>;
```
Just FYI, [`sensor`](https://github.com/zephyrproject-rtos/zephyr/blob/86af9c4b0766842aa97b5b9bead12c71ca5d69ed/drivers/video/video_stm32_dcmi.c#L479) could eventually be swapped for [`source`](https://github.com/zephyrproject-rtos/zephyr/pull/72623#discussion_r1598630537).
Nothing to do here now though!

uLipe replied:
Is there any diference in that sense? Because I blindly copied this section from other examples like provided by NXP and the own binding from ST DCMI.

josuah replied:
Sorry, I meant that some day, the word "sensor" could be replaced by "source" *in a future pull request*.
This is just some "news" about evolution of the API. Here the driver uses "sensor", so no way to use something else for now.

josuah commented on boards/arduino/nicla_vision/camera_ext_clock.c:
```diff
@@ -0,0 +1,43 @@
+#include <zephyr/device.h>
+#include <zephyr/drivers/clock_control.h>
+
+#define LOG_LEVEL CONFIG_CLOCK_CONTROL_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(camera_ext_clock);
```
I could be wrong, but I think this is the older logging declaration format, with a new format [introduced](https://github.com/zephyrproject-rtos/zephyr/commit/5e346812aca93570c544042c99122a059b35e48b) allowing `<zephyr/logging/log.h>` to be a normal header. you could go for the new style:

```suggestion
#include <zephyr/logging/log.h>

LOG_MODULE_REGISTER(camera_ext_clock, CONFIG_CLOCK_CONTROL_LOG_LEVEL);
```

uLipe replied:
I think you are right, I picked this file from an old board init snippet.

josuah reviewed (commented):
(comment only as it is a draft)
I am learning a bit more every review :)

uLipe committed:
drivers: video: gc2145: add resolution support

Add VGA and QVGA resolution support to enable
usage of gc2145 sensor by smaller ram capable devices.

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>

uLipe committed:
samples: drivers: video: add arduino nicla vision

As supported board to the video capture sample.

Signed-off-by: Felipe Neves <ryukokki.felipe@gmail.com>

uLipe commented:
Just ran clang-format again on the GC2145 driver

kartben commented on drivers/video/gc2145.c:
```diff
@@ -689,22 +689,25 @@ struct gc2145_data {
 	{0},
 };
 
+enum vga_resultions {
```
typo :)

kartben commented on drivers/video/gc2145.c:
```diff
@@ -901,28 +904,59 @@ static int gc2145_set_output_format(const struct device *dev, int output_format)
+		win_h = h * 3;
+		break;
+	case VGA_RESOLUTION:
+		win_w = w * 2;
+		win_h = h * 2;
```
This is she same as VGA so probably an oversight?

josuah commented on samples/drivers/video/capture/README.rst:
```diff
@@ -24,6 +24,9 @@ Supported camera modules on some i.MX RT boards can be found below.
 
+Also Arduino Nicla Vision can be used in this sample as capture device, in that case
+The user can transfer the captured frames through on board USB.
+
```
In case you also wanted a link to the board, you can use a reference to [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arduino/nicla_vision/doc/index.rst?plain=1#L1)
```suggestion
- :ref:`arduino_nicla_vision_board`
```

uLipe replied:
Hmm I was sure I tested using that way, and the documentation build has  been failed.

josuah commented on samples/drivers/video/capture/boards/arduino_nicla_vision_stm32h747xx_m7.conf:
```diff
@@ -0,0 +1,15 @@
+
+# Arduino Nicla Vision has easy access to the USB console
+# So use it to check the log messages
+CONFIG_USB_DEVICE_STACK=y
```
Is that controller not supported by `device_next` yet? Just `device`?
Not that it's blocking though. Just one more thing to convert some day.

uLipe replied:
I did not test it, the only thing different with Nicla USB is the fact that it uses an ULPI phy to achieve high speeds.

Since it worked out of the box with the legacy USB device stack, I just kept this as default.

uLipe replied:
Just added, I was adding a unnecessary underscore before.

josuah commented on samples/drivers/video/capture/boards/arduino_nicla_vision_stm32h747xx_m7.overlay:
```diff
@@ -0,0 +1,19 @@
+	chosen {
+		zephyr,console = &cdc_acm_uart0;
+		zephyr,shell-uart = &cdc_acm_uart0;
+		zephyr,cdc-acm-uart0 = &cdc_acm_uart0;
```
Do you know where `zephyr,cdc-acm-uart0` chosen node is used? I cannot find it anywhere.
If nowhere, the could be removed.
I also see it here: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arduino/portenta_h7/arduino_portenta_h7_stm32h747xx_m7.dts#L20 but maybe it is a typo?

uLipe replied:
i get the reference from the Arduino Opta, which defaults its console to USB.

josuah replied:
In my understanding, this has no effect and can be left out, but does not harm either. Maybe some external code relies on this for the other boards it came from.

josuah reviewed (dismissed):
Some feedback left above, which does not affect the functionality.

pillo79 commented on drivers/video/gc2145.c:
```diff
@@ -689,22 +689,25 @@ struct gc2145_data {
 	{0},
 };
 
+enum vga_resolutions {
```
Is there a sensor-specific reason for `vga_` here? Can this (and the variables below) be simply `resolution`? 
Also, if `fmts` is linked with this enum, you could define the enum before the const array and explicitly link the index with it.

pillo79 reviewed (commented):
This looks great! Not an expert here so these are more questions than actual comments, please bear with me.

Also, FYI - even when replying on your own PR, you can "start a review" on GitHub by going to the Files tab. This greatly reduces the number of email messages to other reviewers :slightly_smiling_face:

uLipe commented on drivers/video/gc2145.c:
```diff
@@ -1105,11 +1139,12 @@ static int gc2145_init(const struct device *dev)
-	fmt.pitch = UXGA_HSIZE * 2;
+	fmt.width = fmts[QVGA_RESOLUTION].width_min;
+	fmt.height = fmts[QVGA_RESOLUTION].height_min;
+	fmt.pitch = fmts[QVGA_RESOLUTION].width_min * 2;
```
It should not be a problem, the pitch should be provided by the caller, and then it gets copied to the driver data when setting the resolution.

josuah replied:
[here](https://github.com/zephyrproject-rtos/zephyr/pull/78422/files#diff-83d2e8ab89cc9ed3403bffe5472f4c642b331b0d887612c3bdc8696089e82b3aR1043) I suppose
```c
drv_data->fmt = *fmt;
```

But it looks ignored altogether by the driver: no custom pitch available it seems.

uLipe replied:
Yes, this line would act like a memcpy from the input parameter to the driver data struct, in the sensor driver the pitch itself is not used in runtime, but in the default state it does, because most of video applications may fetch the default format via video_get_caps, and the pitch is used to calculate the amount of space in the video buffer.

uLipe commented on drivers/video/gc2145.c:
```diff
@@ -901,28 +904,59 @@ static int gc2145_set_output_format(const struct device *dev, int output_format)
+	uint16_t win_y;
+
+	/* Add the subsampling factor depending on resolution */
+	switch (vga_res) {
```
I did not get it, could you please clarify?  The resolution set function was something that came from the original Arducam Library, so I did not looked into it with a optimization vision. :)

josuah commented on boards/arduino/nicla_vision/arduino_nicla_vision_stm32h747xx_m7.dts:
```diff
@@ -192,3 +186,81 @@ zephyr_udc0: &usbotg_hs {
+		status = "okay";
+
+		port {
+			gc2145_ep_out: endpoint {
+				remote-endpoint = <&dcmi_ep_in>;
+			};
```
FYI only, there is currently a bug somewhere in the python wrappers around the devicetree to support circular phandle references.

- https://github.com/zephyrproject-rtos/zephyr/issues/57708
- https://github.com/zephyrproject-rtos/zephyr/pull/74415 (workaround)

This does not affect this PR: the workaround is not merged yet.

pillo79 commented on drivers/video/gc2145.c:
```diff
@@ -901,28 +905,65 @@ static int gc2145_set_output_format(const struct device *dev, int output_format)
-	uint16_t y = (((win_h / r_ratio) - UXGA_VSIZE) / 2);
-	uint16_t win_x = ((UXGA_HSIZE - win_w) / 2);
-	uint16_t win_y = ((UXGA_VSIZE - win_h) / 2);
+	uint16_t w = fmts[res].width_min;
+	uint16_t h = fmts[res].height_min;
```
Looks like an oversight - there's already a check and a safe assignment below.
```suggestion
	uint16_t w;
	uint16_t h;
```

pillo79 commented on drivers/video/gc2145.c:
```diff
@@ -901,28 +905,65 @@ static int gc2145_set_output_format(const struct device *dev, int output_format)
+	h = fmts[res].height_min;
+
+	/* Add the subsampling factor depending on resolution */
+	switch (res) {
+	case QVGA_RESOLUTION:
+		win_w = w * 3;
+		win_h = h * 3;
+		break;
+	case VGA_RESOLUTION:
+		win_w = w * 2;
+		win_h = h * 2;
+		break;
+	case UXGA_RESOLUTION:
+		win_w = w;
+		win_h = h;
+		break;
+	default:
+		LOG_ERR("Unsupported resolution %d", res);
+		return -EIO;
+	};
+
+	/* Calculates the window boundaries to obtain the desired resolution */
+	c_ratio = win_w / w;
+	r_ratio = win_h / h;
+	x = (((win_w / c_ratio) - w) / 2);
+	y = (((win_h / r_ratio) - h) / 2);
```
IIUC the cases set `win_w` by multiplying `w` with a hardcoded number. Then, just after the `switch`, `c_ratio` values are calculated by dividing `win_w` by `w`. That's the hardcoded value.. :slightly_smiling_face: 
But this whole block looks fishy to me: `x` and `y` end up always zero (`w` - `w`).  
I suspect this logic originally dealt with rounding issues, at a time when `win_w` was unrelated to `w`. Or I am missing something...

uLipe replied:
Got your point, it seems I'm looking too often to this code that I can't see more the obvious, just fixed.

On the other hand, some of the logic is also imported for the zoom handling, on the Arducam driver, the set resolution original takes also the zoom factor as variable, in that sense I'm discarding it since there is no zoom, much like the original code does, and I also removed the zoom logic, bacause it will not be supported yet (I need to check if zoom could be supported as standard video4zephyr control or if the API would need some kind of extension.

Please let me know if the current fix is enough for now.

uLipe commented:
One more round folks, we are almost there :)

pillo79 reviewed (approved):
Thanks!
